Extend late restartable data

loganharbour opened the pull request:
refs #23352

Still need to:

- [ ] Add proper checking when lazily declaring data and making sure it is loaded later

loganharbour committed:
Expand capability to get late data

refs #23352

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1798920/) on 3e5b727 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25625/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25625/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3923fccb00ea7c022ebc12e59b567191c1693d96`

loganharbour committed:
Use new read/write API in stochastic_tools

refs #23352

loganharbour committed:
Apply clang-format

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1798979/) on 1ed5703 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25625/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25625/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3923fccb00ea7c022ebc12e59b567191c1693d96`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1799046/) on 8ab640f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25625/site)

This comment will be updated on new commits.

loganharbour committed:
Loosen re-registration check

refs #23352

loganharbour committed:
Correct spelling

loganharbour committed:
Update error message

refs #23352

loganharbour committed:
Add static assertion for late loading of data

loganharbour committed:
Simplify restartableName, remove unused methods

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1799274/) on 8d6562c wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>3923fc</th>
      <th colspan=4>#25625 8d6562</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.69%</td>
      <td>85.69%</td>
      <td>-0.00%</td>
      <td>92.03%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>91789</td>
      <td>91829</td>
      <td>+40</td>
      <td>127</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15330</td>
      <td>15340</td>
      <td>+10</td>
      <td>11</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25625/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25625/coverage/framework)


## Modules coverage

### Reactor

<table>
  <thead>
    <tr>
      <th></th>
      <th>3923fc</th>
      <th colspan=4>#25625 8d6562</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.78%</td>
      <td>95.77%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4920</td>
      <td>4919</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>217</td>
      <td>217</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25625/coverage_diff/reactor)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25625/coverage/reactor)

### Stochastic tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>3923fc</th>
      <th colspan=4>#25625 8d6562</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>91.23%</td>
      <td>91.23%</td>
      <td>-0.01%</td>
      <td>80.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7669</td>
      <td>7663</td>
      <td>-6</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>737</td>
      <td>737</td>
      <td>-</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25625/coverage_diff/stochastic_tools)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25625/coverage/stochastic_tools)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25625/coverage/xfem)

</details>

## Warnings

- `stochastic_tools` new line coverage rate 80.00% is less than the suggested 90.0%



This comment will be updated on new commits.

loganharbour committed:
Simplify meta data registration

loganharbour committed:
Add late loading capability to Restartable

refs #23352

loganharbour committed:
Allow for an empty system and restartable name

loganharbour committed:
Add more useful assertion message

loganharbour committed:
Add error, don't load unneeded data

loganharbour committed:
Un-constify getting data

This cannot be const anymore because we could potentially
load late data, which isn't const

loganharbour committed:
Add more useful error when using a meta map

loganharbour committed:
Use restored state to help determine if data exists

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1852167/) on 3bad960 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25625/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25625/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f1b7bfa6b92d76ae14f4944e0f96561980e933d6`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1852167/) on 3bad960 wanted to post the following:

The following commit(s) contain fixups:

- c3e1e9c377: fixup! 2027c3253d0000132f0cde32f363433a99e601b4
- 4439546727: fixup! 085ab72f6eb517d8b172bd9ce67ffa65b7e383c3
- c486288849: fixup! 085ab72f6eb517d8b172bd9ce67ffa65b7e383c3

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1852167/) on 3bad960 wanted to post the following:

The following commit(s) contain fixups:

- c3e1e9c377: fixup! 2027c3253d0000132f0cde32f363433a99e601b4
- 4439546727: fixup! 085ab72f6eb517d8b172bd9ce67ffa65b7e383c3
- c486288849: fixup! 085ab72f6eb517d8b172bd9ce67ffa65b7e383c3

loganharbour committed:
Apply clang-format

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1861259/) on d42608a wanted to post the following:

The following commit(s) contain fixups:

- c3e1e9c377: fixup! 2027c3253d0000132f0cde32f363433a99e601b4
- 4439546727: fixup! 085ab72f6eb517d8b172bd9ce67ffa65b7e383c3
- c486288849: fixup! 085ab72f6eb517d8b172bd9ce67ffa65b7e383c3

loganharbour committed:
Finalize the logic for checking for a restartable value

loganharbour committed:
Improve loading without a store late test

loganharbour committed:
Stochastic tools fixes for restart/recover changes

loganharbour committed:
Correct context, add hasRestored checks

refs #23352

loganharbour committed:
Make non-unity builds great again

loganharbour committed:
Skip recover tests here

These are already testing recover

refs #23352

loganharbour committed:
Re-add deprecated name so that we don't break griffin

refs #23352

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1869328/) on 6c95998 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)

loganharbour committed:
Add support for multiple streams (unused)

refs #23352

loganharbour committed:
Correct assertion and unit test
